src/Analyzers/CSharp/Analyzers/UsePatternCombinators/CSharpUsePatternCombinatorsDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,144 @@
+                SyntaxKind.ArrowExpressionClause,
+                SyntaxKind.SimpleAssignmentExpression,
+                SyntaxKind.SimpleLambdaExpression,
+                SyntaxKind.ParenthesizedLambdaExpression,
+                SyntaxKind.Argument);
+
+        private static ExpressionSyntax? GetExpression(SyntaxNode node)
+            => node switch
+            {
+                SwitchExpressionArmSyntax n => n.Expression,
+                ConditionalExpressionSyntax n => n.Condition,
+                ForStatementSyntax n => n.Condition,
+                EqualsValueClauseSyntax n => n.Value,
+                IfStatementSyntax n => n.Condition,
+                WhenClauseSyntax n => n.Condition,
+                WhileStatementSyntax n => n.Condition,
+                DoStatementSyntax n => n.Condition,
+                ReturnStatementSyntax n => n.Expression,
+                YieldStatementSyntax n => n.Expression,
+                ArrowExpressionClauseSyntax n => n.Expression,
+                AssignmentExpressionSyntax n => n.Right,
+                LambdaExpressionSyntax n => n.ExpressionBody,
+                ArgumentSyntax n when n.GetRefKind() == RefKind.None => n.Expression,
+                _ => null,
+            };
+
+        private void AnalyzeNode(SyntaxNodeAnalysisContext context)
+        {
+            var parentNode = context.Node;
+            var syntaxTree = parentNode.SyntaxTree;
+            var cancellationToken = context.CancellationToken;
+
+            if (!((CSharpParseOptions)syntaxTree.Options).LanguageVersion.IsCSharp9OrAbove())
+                return;
+
+            var styleOption = context.Options.GetOption(CSharpCodeStyleOptions.PreferPatternMatching, syntaxTree, cancellationToken);
+            if (!styleOption.Value)
+                return;
+
+            var expression = GetExpression(parentNode);
+            if (expression is null)
+                return;
+
+            var operation = context.SemanticModel.GetOperation(expression, cancellationToken);
+            if (operation is null)
+                return;
+
+            var pattern = CSharpUsePatternCombinatorsAnalyzer.Analyze(operation);
+            if (pattern is null)
+                return;
+
+            // Avoid rewriting trivial patterns, such as a single relational or a constant pattern.
+            if (IsTrivial(pattern))
+                return;
+
+            // C# 9.0 does not support pattern variables under `not` and `or` combinators.
+            if (HasIllegalPatternVariables(pattern))
+                return;
+
+            context.ReportDiagnostic(DiagnosticHelper.Create(
+                Descriptor,
+                expression.GetLocation(),
+                styleOption.Notification.Severity,
+                additionalLocations: null,
+                properties: null));
+        }
+
+        private static bool HasIllegalPatternVariables(AnalyzedPattern pattern, bool permitDesignations = true)
```

Comment:
there's a large pity here that we effectively have the same check in two different features written two different ways.  (same as the pity in having written the 'negate' function twice) :-/

Reply:
I have some ideas to extend both if-to-switch and switch-stmt-to-expr that could share a lot with this feature (and some new). I can revise once I get there.